prow/cmd/peribolos/main.go

```diff
@@ -103,21 +128,132 @@ func main() {
 	c.Throttle(300, 100) // 300 hourly tokens, bursts of 100
 
 	for name, orgcfg := range cfg.Orgs {
-		if err := configureOrg(c, name, orgcfg); err != nil {
-			log.Fatalf("Configuration failed: %v", err)
+		if err := configureOrg(o, c, name, orgcfg); err != nil {
+			logrus.Fatalf("Configuration failed: %v", err)
 		}
 	}
 }
 
-func configureOrg(client *github.Client, orgName string, orgConfig org.Config) error {
+type configureOrgMembersClient interface {
+	BotName() (string, error)
+	ListOrgMembers(org, role string) ([]github.TeamMember, error)
+	RemoveOrgMembership(org, user string) error
+	UpdateOrgMembership(org, user string, admin bool) (*github.OrgMembership, error)
+}
+
+func configureOrgMembers(opt options, client configureOrgMembersClient, orgName string, orgConfig org.Config) error {
+	// Get desired state
+	wantAdmins := sets.NewString(orgConfig.Admins...)
+	wantMembers := sets.NewString(orgConfig.Members...)
+
+	// Sanity desired state
+	if n := len(wantAdmins); n < opt.minAdmins {
+		return fmt.Errorf("%s must specify at least %d admins, only found %d", orgName, opt.minAdmins, n)
+	}
+	var missing []string
+	for _, r := range opt.requiredAdmins.Strings() {
+		if !wantAdmins.Has(r) {
+			missing = append(missing, r)
+		}
+	}
+	if len(missing) > 0 {
+		return fmt.Errorf("%s must specify %v as admins, missing %v", orgName, opt.requiredAdmins, missing)
+	}
+	if opt.requireSelf {
+		if me, err := client.BotName(); err != nil {
+			return fmt.Errorf("cannot determine user making requests for %s: %v", opt.token, err)
+		} else if !wantAdmins.Has(me) {
+			return fmt.Errorf("authenticated user %s is not an admin of %s", me, orgName)
+		}
+	}
+
+	// Get current state
+	haveAdmins := sets.String{}
+	haveMembers := sets.String{}
+	if ms, err := client.ListOrgMembers(orgName, "admin"); err != nil {
+		return fmt.Errorf("failed to list %s admins: %v", orgName, err)
+	} else {
+		for _, m := range ms {
+			haveAdmins.Insert(m.Login)
+		}
+	}
+
+	if ms, err := client.ListOrgMembers(orgName, "member"); err != nil {
+		return fmt.Errorf("failed to list %s members: %v", orgName, err)
+	} else {
+		for _, m := range ms {
+			haveMembers.Insert(m.Login)
+		}
+	}
+
+	// Ensure desired state is sane
+	both := wantAdmins.Intersection(wantMembers)
+	if n := len(both); n > 0 {
+		return fmt.Errorf("%d users are both a member and admin: %v", n, both)
+	}
+
+	// Figure out who to remove
+	have := haveMembers.Union(haveAdmins)
+	want := wantMembers.Union(wantAdmins)
+	remove := have.Difference(want)
+
+	// Figure out who to invite and/or reconfigure
+	makeMember := wantMembers.Difference(haveMembers)
+	makeAdmin := wantAdmins.Difference(haveAdmins)
+
+	// Sanity check changes
+	if d := float64(len(remove)) / float64(len(have)); d > opt.maximumDelta {
+		return fmt.Errorf("cannot delete %d memberships or %.3f of %s (exceeds limit of %.3f)", len(remove), d, orgName, opt.maximumDelta)
```

Comment:
What is the use of the `opt.maximumDelta`? To bound API calls? Why does it scale with the size of the org being operated on?

Reply:
This is a common pattern for dealing with large numbers. Each user's bill will vary, but ensure total revenue hasn't done something absurd like halved or doubled in a day; the exact amount of traffic to any given page will vary and total traffic will vary over time, but alert if total traffic to the site changes dramatically, etc.

Expected, important scenarios to support:
a) Submitting a PR to adding/removing a person or two (removals are under maximumDelta)
b) Seed config for a new org, potentially adding a bunch of people (adding a bunch of people is fine)

On the other hand, deleting half the org's admins in a single run feels more like a software problem, disgruntled member, or some other exceptional situation that should require manual review (aka running the job with `--maximum-removal-delta=higher-value-like-1.0`.